### PR TITLE
Add coarse-to-fine and second order inference for coref. Also a new coref config

### PR DESCRIPTION
**Pull request by ZhaofengWu**
Based on https://arxiv.org/pdf/1804.05392.pdf. 79.17 dev 78.87 test. URL: https://storage.googleapis.com/allennlp-public-models/coref-spanbert-large-2020.02.27.tar.gz

SpanBERT reports 79.6 on test. According to the e2e-coref paper, speaker and genre metadata, which we aren't using, contributes 1.4 F1. So I assume that's where the difference is coming from.

### CODE REVIEW

File: allennlp/models/coreference_resolution/coref.py [hunk 1]
```diff
@@ -21,10 +21,10 @@
 @Model.register("coref")
 class CoreferenceResolver(Model):
     """
-    This `Model` implements the coreference resolution model described "End-to-end Neural
-    Coreference Resolution"
-    <https://www.semanticscholar.org/paper/End-to-end-Neural-Coreference-Resolution-Lee-He/3f2114893dc44eacac951f148fbff142ca200e83>
-    by Lee et al., 2017.
+    This `Model` implements the coreference resolution model described in
+    "Higher-order Coreference Resolution with Coarse-to-fine Inference"
```
Comment (DeNeutoy) [inline allennlp/models/coreference_resolution/coref.py]: Can you use a markdown link?

File: allennlp/models/coreference_resolution/coref.py [hunk 2]
```diff
@@ -53,6 +53,10 @@ class CoreferenceResolver(Model):
         spans we retain with respect to the number of words in the document.
     max_antecedents: int, required.
         For each mention which survives the pruning stage, we consider this many antecedents.
+    coarse_to_fine: bool, optional (default = False)
+        Whether or not to apply the coarse-to-fine filtering.
+    num_coref_layers: int, optional (default = 1)
```
Comment (DeNeutoy) [inline allennlp/models/coreference_resolution/coref.py]: `num_span_refinement_layers` might be a better name, given your description 😄

File: allennlp/models/coreference_resolution/coref.py [hunk 3]
```diff
@@ -472,15 +525,185 @@ def _generate_valid_antecedents(
...
+        return valid_antecedent_indices, valid_antecedent_offsets, valid_antecedent_mask
+
+    def _distance_pruning(
+        self,
+        top_span_embeddings: torch.FloatTensor,
+        top_span_mention_scores: torch.FloatTensor,
+        max_antecedents: int,
+    ) -> Tuple[torch.FloatTensor, torch.BoolTensor, torch.LongTensor, torch.LongTensor]:
+        """
+        Generates antecedents for each span and prunes down to `max_antecedents`.
```
Comment (DeNeutoy) [inline allennlp/models/coreference_resolution/coref.py]: Can you add a more comprehensive description of the pruning methods here and below?

File: allennlp/models/coreference_resolution/coref.py [hunk 4]
```diff
@@ -472,15 +525,185 @@ def _generate_valid_antecedents(
...
+        """
+        batch_size, num_spans_to_keep = top_span_embeddings.size()[:2]
+        device = util.get_device_of(top_span_embeddings)
+
+        # Shape: (1, num_spans_to_keep, num_spans_to_keep)
+        _, _, valid_antecedent_mask = self._generate_valid_antecedents(
+            num_spans_to_keep, num_spans_to_keep, device
+        )
+
+        # Shape: (batch_size, num_spans_to_keep, num_spans_to_keep); broadcast op
```
Comment (DeNeutoy) [inline allennlp/models/coreference_resolution/coref.py]: Can you make this section a bit easier to read? I might write it like:
```
	mention_one_score = top_span_mention_scores.unsqueeze(0)
	mention_two_score = top_span_mention_scores.unsqueeze(1)
	coarse_to_fine_weights = self._coarse2fine_scorer(top_span_embeddings).transpose(1, 2)
	coarse_to_fine_scores = torch.matmul(top_span_embeddings, coarse_to_fine_weights)		
	
	partial_antecedent_scores = coarse_to_fine_scores + mention_one_score + mention_two_score
```

File: allennlp/models/coreference_resolution/coref.py [hunk 5]
```diff
@@ -472,15 +525,185 @@ def _generate_valid_antecedents(
...
+        ) + top_span_mention_scores.unsqueeze(2)
+        partial_antecedent_scores += torch.matmul(
+            top_span_embeddings, self._coarse2fine_scorer(top_span_embeddings).transpose(1, 2)
+        )
+
+        # Shape: (batch_size, num_spans_to_keep, num_spans_to_keep); broadcast op
+        span_pair_mask = top_span_mask.unsqueeze(-1) & valid_antecedent_mask
+
+        # Shape:
+        # (batch_size, num_spans_to_keep, max_antecedents) * 3
```
Comment (DeNeutoy) [inline allennlp/models/coreference_resolution/coref.py]: I was confused by your `* 3`, can you just say "all of size: "

File: allennlp/models/coreference_resolution/coref.py [hunk 6]
```diff
@@ -505,7 +728,7 @@ def _compute_span_pair_embeddings(
             (batch_size, num_spans_to_keep, max_antecedents, embedding_size).
         antecedent_offsets : `torch.IntTensor`, required.
             The offsets between each top span and its antecedent spans in terms
-            of spans we are considering. Has shape (1, max_antecedents).
+            of spans we are considering. Has shape (batch_size, num_spans_to_keep, max_antecedents).
```
Comment (DeNeutoy) [inline allennlp/models/coreference_resolution/coref.py]: I don't understand this change - was this comment wrong? Why do we need batch size here again? Is it because the span pair embeddings are no longer symmetric or something?

File: allennlp/models/coreference_resolution/coref.py [hunk 7]
```diff
@@ -605,17 +815,23 @@ def _compute_coreference_scores(
...
-            (batch_size, num_spans_to_keep, max_antecedents, encoding_dim)
-        top_span_mention_scores : `torch.FloatTensor`, required.
-            Mention scores for every span. Has shape
+        top_span_embeddings : `torch.FloatTensor`, required.
+            Embedding representations of the kept spans. Has shape
+            (batch_size, num_spans_to_keep, embedding_size)
+        top_antecedent_embeddings: `torch.FloatTensor`, required.
+            The embeddings of antecedents for each span candidate. Has shape
+            (batch_size, num_spans_to_keep, max_antecedents, embedding_size)
+        top_partial_antecedents_scores : `torch.FloatTensor`, required.
```
Comment (DeNeutoy) [inline allennlp/models/coreference_resolution/coref.py]: Can you explain why this is partial, it's not clear right now what it is missing

File: allennlp/tests/models/coreference_resolution/coref_test.py [hunk 8]
```diff
@@ -12,7 +12,24 @@ def setUp(self):
         )
 
     def test_coref_model_can_train_save_and_load(self):
-        self.ensure_model_can_train_save_and_load(self.param_file)
+        for coarse_to_fine in (True, False):
+            for num_coref_layes in (1, 3):
+                self._test_coref_model_can_train_save_and_load(coarse_to_fine, num_coref_layes)
```
Comment (DeNeutoy) [inline allennlp/tests/models/coreference_resolution/coref_test.py]: \s\num_coref_layes\num_coref_layers

File: allennlp/models/coreference_resolution/coref.py [hunk 9]
```diff
@@ -361,12 +412,14 @@ def make_output_human_readable(self, output_dict: Dict[str, torch.Tensor]):
         # A tensor of shape (num_spans_to_keep, max_antecedents), representing the indices
         # of the predicted antecedents with respect to the 2nd dimension of `batch_top_spans`
         # for each antecedent we considered.
-        antecedent_indices = output_dict["antecedent_indices"].detach().cpu()
+        batch_antecedent_indices = output_dict["antecedent_indices"].detach().cpu()
```
Comment (DeNeutoy) [inline allennlp/models/coreference_resolution/coref.py]: Ah, I think I understand why we need the batch dim - you want to have different possible numbers of antecedent indices per batch?

File: training_config/coref_spanbert_large.jsonnet [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+    "antecedent_feedforward": {
+        "input_dim": span_pair_embedding_dim,
+        "num_layers": 2,
+        "hidden_dims": 1500,
+        "activations": "relu",
+        "dropout": 0.3
+    },
+    "initializer": {
+      "regexes": [
+        [".*linear_layers.*weight", {"type": "xavier_normal"}],
```
Comment (DeNeutoy) [inline training_config/coref_spanbert_large.jsonnet]: You may want to initialize the gated sum module for the refinement, just as an FYI.

**Comment by ZhaofengWu** [comment]
@DeNeutoy When `max_antecedents` is 100, the delta with respect to coarse-to-fine is 2.24 F1 on dev. We are using small `max_antecedents` values due to the large models using more memory, so the difference is a lot more significant than in the original paper. I had to tune `max_antecedents` further down to 50 to avoid OOM when incorporating 2nd order inference, but that's fine because 50 was also the value used by the SpanBERT people. I didn't rerun the experiments, but I image coarse-to-fine helps even more at 50 antecedents.

File: allennlp/models/coreference_resolution/coref.py (hunk 6 above)
Comment (ZhaofengWu) [reply allennlp/models/coreference_resolution/coref.py]: Yes, for the non-coarse-to-fine case, we always select the antecedents with the same offsets, i.e. `[max_antecedents, 0)`. But the coarse-to-fine case allows more flexibility so that each span can select arbitrary antecedents with a scoring function.

**Comment by DeNeutoy** [comment]
Awesome that's great 👍 Consider most of my points above superficial in that case/just questions.

File: allennlp/models/coreference_resolution/coref.py (hunk 9 above)
Comment (ZhaofengWu) [reply allennlp/models/coreference_resolution/coref.py]: Not just different numbers of antecedent indices, but really different values of antecedent indices, due to the flexibility of the coarse-to-fine scorer.

File: allennlp/models/coreference_resolution/coref.py (hunk 2 above)
Comment (ZhaofengWu) [reply allennlp/models/coreference_resolution/coref.py]: Yeah my description wasn't 100% accurate either. Updated both.

File: allennlp/models/coreference_resolution/coref.py (hunk 7 above)
Comment (ZhaofengWu) [reply allennlp/models/coreference_resolution/coref.py]: Explained in the doc.

**Comment by rahular** [comment]
Hi @ZhaofengWu, thank you for the awesome work! I want to know which GPU(s) you use to train the bert coref models? A single 12gb GPU is running OOM.

**Comment by ZhaofengWu** [comment]
@rahular I used a 48GB GPU. You can adjust [the config file](https://github.com/allenai/allennlp/blob/master/training_config/coref_spanbert_large.jsonnet) to your GPU spec accordingly, e.g. by using the SpanBERT base model.
